README.md
[![CircleCI](https://circleci.com/gh/jstewmc/php-helpers.svg?style=svg)](https://circleci.com/gh/jstewmc/php-helpers) [![codecov](https://codecov.io/gh/jstewmc/php-helpers/branch/master/graph/badge.svg?token=nO0rvmLWUk)](https://codecov.io/gh/jstewmc/php-helpers)

# PHP Helpers
Static classes to help with PHP strings, arrays, numbers, files, and boolean values.

Static helper classes are nothing new in PHP. In fact, most of these functions have probably been written dozens of times in better libraries than mine. However, I wrote (or copied these functions from the web with credit) when I worked on a project that required as few dependencies as possible, and I figured I would share them.

## Installation

This library requires [PHP 7.4+](https://secure.php.net).

It is multi-platform, and we strive to make it run equally well on Windows, Linux, and OSX.

It should be installed via [Composer](https://getcomposer.org). To do so, add the following line to the `require` section of your `composer.json` file, and run `composer update`:

```javascript
{
   "require": {
       "jstewmc/php-helpers": "^0.2"
   }
}
```

## Usage

Here are examples for the most commonly used functions.

(Please note, I've omitted the requisite `use Jstewmc\PhpHelpers\{Arr, Boolean, Dir, Num, Str};` statements to keep the examples cleaner.)

### Numbers (aka, "Num")

You can use `val()` method to evaluate integers, floats, fractions (e.g., `"1/2"`), mixed numbers (e.g., `"1 1/2"`), comma-separated values (e.g., `"1,000"`), and english-worded numbers (e.g., `"two hundred and fifty-six"`) to their numeric equivalent:

```php
Num::val('1/2');          // returns (float) 0.5
Num::val('1,000');        // returns (int) 1000
Num::val('one hundred');  // returns (int) 100

Num::val('two million, ninety-seven thousand, one hundred and fifty-two');  
// returns (int) 2,097,152
```

You can use the `*To()` methods to round, ceil, or floor a number to the nearest multiple of another number:

```php
Num::roundTo(7, 10);  // returns 10
Num::ceilTo(7, 10);   // returns 10
Num::floorTo(7, 10);  // returns 0
```

You can use the `bound()` method to keep a number greater than or equal to a lower bound, less than or equal to an upper bound, or both:

```php
Num::bound(1, 10);  
// returns 10, because 1, the number, is less than 10, the lower bound

Num::bound(10, 1, 5);
// returns 5, because 10, the number, is greater than 5, the upper bound
```

You can use the `normalize()` method to index a number according to a maximum value:

```php
Num::normalize(1, 10);   // returns 0.1
Num::normalize(5, 10);   // returns 0.5
Num::normalize(10, 10);  // returns 1
```

You can use the `isNumeric()` method to test whether or not a value is a number, including fractions, mixed numbers, and english phrases:

```php
Num::isNumeric(1);             // returns true
Num::isNumeric('1/2');         // returns true
Num::isNumeric('1 1/2');       // returns true
Num::isNumeric('one hundred'); // returns true
Num::isNumeric('foo');         // returns false
```

You can use the `isInt()` method to test whether or not a number or string is an integer:

```php
Num::isInt('1,000');  // returns true
Num::isInt(1000);     // returns true
```

You can use the `isId()` method to test whether or not a number is a valid database identifier (i.e., a positive integer, optionally of the correct datatype size):

```php
Num::isId('foo');           // returns false
Num::isId(1.5);             // returns false
Num::isId(1);               // returns true
Num::isId(1, 'tinyint');    // returns true
Num::isId(999, 'tinyint');  // returns false (too big)
```

You can use the `isZero()` method to test whether or not a value is zero (in loosely-typed languages like PHP, zero can be many things):

```php
Num::isZero(0);      // returns true
Num::isZero('0');    // returns true
Num::isZero(false);  // returns false
```

You can use the `almostEqual()` method to test whether or not two floats are "equal" (because of the way floats are stored in memory, you shouldn't compare them directly using the `==` or `===` operators):

```php
Num::almostEqual(0.2, 0.2);  // returns true
Num::almostEqual(0.2, 0.3);  // returns false
```

### Strings (aka, "Str")

You can use `rand()` to generate a random string of a given length, optionally with specific latin character sets (allowed character sets are `'lower'`, `'upper'`, `'alpha'` (a shortcut for `'lower' + 'upper'`), `number`, or `symbol`):

``` php
Str::rand(8);                       // returns a string like '9#hb%Fv3'
Str::rand(8, ['upper', 'number']);  // returns a string like 'P9K7HG32'
```

You can use `password()` to generate a random string, optionally with the minimum number of characters that must be present from latin character sets:

```php
Str::password(8);
// returns a string like 'jNb^3#L@'

Str::password(8, ['upper' => 8]);   
// returns a string like 'NBDRATCV', with exactly eight upper-case characters

Str::password(16, ['number' => 8]);  
// returns a string like '*9f8F6b4F3f1:0/9', with at least eight numbers
```

You can use `truncate()` to neatly cut a string at or near the desired length (by default, the break character is the space character (`' '`) and the padding is an ellipsis (`'...'`)):

```php
Str::truncate('Lorem ipsum inum', 10);  
// returns 'Lorem...', because the "u" in "ipsum" is the 10th character and
// the space after "Lorem" is the closest break character

Str::truncate('Lorem ipsum inum', 15);  
// returns 'Lorem ipsum...', because the "u" in "inum" is the 15th character

Str::truncate('Lorem ipsum inum', 99);  
// returns 'Lorem ipsum inum', because the string is shorter than the limit
```

You can use the `*With()` methods to determine whether or not a string starts or ends with a given substring:

```php
Str::endsWith('foo', 'o');     // returns true
Str::iEndsWith('foo', 'O')     // returns true
Str::startsWith('foo', 'f');   // returns true
Str::iStartsWith('foo', 'F');  // returns true
```

You can use the `strtocamelcase()` method to convert a string to camel-case:

```php
Str::strtocamelcase('Hello world');   // returns "helloWorld"
Str::strtocamelcase('H3LLO WORLD!');  // returns "helloWorld"
Str::strtocamelcase('hello_world');   // returns "helloWorld"
```

You can use the `splitOnFirstAlpha()` method to split a string on the first alphabetical character:

```php
Str::splitOnFirstAlpha('123 foo');  // returns ['123', 'foo']
Str::splitOnFirstAlpha('123');      // returns ['123']
Str::splitOnFirstAlpha('foo');      // returns ['foo']
```

You can use the `strtobytes()` method to convert an `.ini`-style byte string (e.g., `'1G'`) to a number:

```php
Str::strtobytes('1K');  // returns 1024
Str::strtobytes('1M');  // returns 1,048,576
```

You can use the `isBool()` method to determine whether or not a string is a "bool-ish" value:

```php
Str::isBool('foo');  // returns false
Str::isBool('yes');  // returns true
```

### Booleans

You can use the `booltostr()` method to convert a boolean value to a string:

``` php
Boolean::booltostr(true, 'yes/no');      // returns 'yes'
Boolean::booltostr(true, 'true/false');  // returns 'true'
Boolean::booltostr(true, 'on/off');      // returns 'on'
```

You can use the `val()` method to evaluate a "bool-ish" value to its boolean equivalent (PHP's native `boolval()` method doesn't support `"yes"`/`"no"` or `"on"`/`"off"` strings):

```php
Boolean::val(true);   // returns true
Boolean::val('on');   // returns true
Boolean::val('yes');  // returns true
```

### Arrays (aka, "Arr")

You can use the `filterByKeyPrefix()` to filter an array by a string prefix:

```php
$array = ['foo' => 1, 'bar' => 2, 'baz' => 3];

Arr::filterByKeyPrefix($array, 'b');  // returns ['bar' => 2, 'baz' => 2]
```

You can use the `filterByKey()` method to filter an array by key using a custom function:

```php
$array = ['foo' => 1, 'bar' => 2, 'baz' => 3];

Arr::filterByKey($a, function ($k) {
	return substr($k, 0, 1) === 'b';  
});
// returns ['bar' => 2, 'baz' => 2]
```

You can use the `sortByField()` method to sort an array of _associative arrays_ in ascending or descending order:

```php
$arrays = [['foo' => 2], ['foo' => 3], ['foo' => 1]];

Arr::sortByField($arrays, 'foo');          
// returns [['foo' => 1], ['foo' => 2], ['foo' => 3]]

Arr::sortByField($arrays, 'foo', 'desc');  
// returns [['foo' => 3], ['foo' => 2], ['foo' => 1]]
```

You can use the `sortByProperty()` or `sortByMethod()` methods to sort an array of _objects_ in ascending or descending order, using a property or method, respectively:

```php
// define a example class (for the purposes of this example, we'll define both a
// public property and a getter method)
class Example
{
	public $property;

	public function __construct(int $value)
	{
		$this->property = $value;
	}

	public function getProperty(): int
	{
		return $this->property;
	}
}

$objects = [new Example(2), new Example(3), new Example(1)];

Arr::sortByProperty($objects, 'foo');  
// returns (in pseudo-code) [{bar: 1}, {bar: 2}, {bar: 3}]

Arr::sortByMethod($objects, 'getBar');  
// returns (in pseudo-code) [{bar: 1}, {bar: 2}, {bar: 3}]
```

You can use the `inArray()` method to search for values in an array using wildcard notation (by default, the wildcard character is the asterisk character (`"*"`)):

```php
$values = ['foo', 'bar', 'baz'];

Arr::inArray($values, 'f*');   // returns true, because of the leading "f" in "foo"
Arr::inArray($values, '*z');   // returns true, because of the trailing "z" in "baz"
Arr::inArray($values, '*a*');  // returns true, because of the "a" in "bar" and "baz"
```

You can use the `diff()` method to determine the [Levenshtein Distance](https://en.wikipedia.org/wiki/Levenshtein_distance), the number of single-element edits required to change one array into another) between two arrays:

```php
$array1 = ['foo', 'bar', 'baz'];
$array2 = ['bar', 'qux'];

$actual = Arr::diff($array1, $array2);

$expected = [
	['value' => 'foo', 'mask' => -1],  // because "foo" should be deleted
	['value' => 'bar', 'mask' => 0],   // because "bar" should be unchanged
	['value' => 'baz', 'mask' => -1],  // because "baz" should be deleted
	['value' => 'qux', 'mask' => 1]    // because "qux" should be added
];

$actual == $expected; // returns true
```

You can use the `permute()` method to calculate an array's permutations (careful, the number of permutations grows as a factorial of the size of the original array):

```php
$array = ['foo', 'bar', 'baz'];

$actual = Arr::permute($array);

$expected = [
	['foo', 'bar', 'baz'],
	['baz', 'foo', 'bar'],
	['bar', 'foo', 'baz'],
	['foo', 'baz', 'bar'],
	['bar', 'baz', 'foo'],
	['baz', 'bar', 'foo']
];

$actual == $expected;  // returns true
```

You can use the `isAssoc()` method to determine whether or not an array is associative (i.e., has a string key):

```php
$array1 = [0 => 'foo', 1 => 'bar'];
$array2 = [0 => 'foo', 1 => 'bar', 'baz' => 'qux'];

Arr::isAssoc($array1);  // returns false
Arr::isAssoc($array2);  // returns true, because there is a string key
```

You can use the `isEmpty()` method to determine whether or not a key exists in an array with a non-empty value:

```php
$values = ['foo' => null, 'bar' => [], 'baz' => 1];

Arr::isEmpty('qux', $values);  
// returns true, because the key "qux" does not exist

Arr::isEmpty('foo', $values);  
// returns true, because the value of key "foo" is null

Arr::isEmpty('bar', $values);  
// returns true, because the value of key "bar" is empty array

Arr::isEmpty('baz', $values);  
// returns false, because the value of key "baz" is not empty
```

You can use the `keyStringReplace()` method to replace substrings in array keys:

```php
$array = ['foo' => 'bar', 'baz' => 'qux'];

Arr::keyStringReplace('f', 'g', $a);
// returns ['goo' => 'bar', 'baz' => 'qux'], because "f" was replaced with "g"

Arr::keyStringReplace('f', '', $a);   
// returns ['oo' => 'bar', 'baz' => 'qux'], because "f" was replaced with ""
```

### Directories (aka, "Dir")

You can use the `copy()` method to duplicate a non-empty directory (PHP's native `copy()` method will not work with non-empty directories):

```php
$source = dirname(__FILE__).'/foo';
$destination = dirname(__FILE__).'/bar';

Dir::copy($source, $destination);  // returns true
```

You can use the `remove()` method to delete a non-empty directory (PHP's native `rmdir()` method will not work with non-empty directories):

```php
$directory = dirname(__FILE__).'/foo';

// The $container verifies you don't delete a directory outside of the target
// area accidentally. The path of the directory to be deleted MUST start with
// this path.
$container = dirname(__FILE__);  

Dir::remove($directory, $container);  // returns true
```

You can use the `abs2rel()` method to convert an absolute path name to a relative one:

```php
Dir::abs2rel('/path/to/foo/bar/baz', '/path/to');  // returns "foo/bar/baz"
Dir::abs2rel('/path/to/foo/bar/baz', '/path/to/foo');  // returns "bar/baz"
```

## License

This library is released under the [MIT License](LICENSE).

## Contributing

[Contributions](contributing.md) are welcome!


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/), and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [0.2.0] - 2021-09-30

### Added

- Added [slevomat/coding-standard](https://github.com/slevomat/coding-standard) to enforce coding standards.
- Added [roave/security-advisories](https://github.com/Roave/SecurityAdvisories) to exclude dependencies with known vulnerabilities.
- Added continuous integration with CircleCI.

### Changed

- Renamed `Bool` to `Boolean` for PHP7 compatibility ([#2](https://github.com/jstewmc/php-helpers/issues/2)).
- Modernized all helpers (e.g., argument type hints, return type hints, guard clauses, etc).
- Updated `PHPUnit` from version 4 to 9.
- Moved tests for `Dir` from concrete files and folders to virtual file system using [bovigo/vfsStream](https://github.com/bovigo/vfsStream).
- Ignored `composer.phar`, `PHPUnit` files, and some system files like `.DS_Store`.
- Updated `Num::isNumeric()` to support english phrases like `two hundred and fifty-six`.

### Removed

- Lots and lots of unnecessary comments :)
- Removed `Dir::cp()` and `Dir::rm()` aliases (they would not have worked anyway).

## [0.1.1] - 2014-10-18

### Added

- Update `Num::val()` to handle strings like "first", "2nd", and "one million four thousand and forty-seven".

## [0.1.0] - 2014-08-27

The initial release.


src/Str.php
<?php

namespace Jstewmc\PhpHelpers;

/**
 * The string (aka, "str") class
 */
class Str
{
    /**
     * Returns true if $haystack ends with $needle (case-sensitive)
     *
     * For example:
     *
     *     Str::endsWith('foobar', 'bar');  // returns true
     *     Str::endsWith('foobar', 'baz');  // returns false
     *     Str::endsWith('foobar', 'BAR');  // returns false
     *     Str::endsWith('foobar', '');     // returns false
     *     Str::endsWith('', 'foobar');     // returns false
     *
     * @param  string  $haystack  the string to search
     * @param  string  $needle    the substring to search for
     *
     * @return  bool  true if $haystack ends with $needle
     *
     * @see  iEndsWith()  case-insensitive version
     * @see  http://stackoverflow.com/a/834355  MrHus' answer to "startsWith()
     *    and endsWith() functions in PHP" on StackOverflow
     */
    public static function endsWith(string $haystack, string $needle): bool
    {
        if (strlen($haystack) > 0 && strlen($needle) > 0) {
            return substr($haystack, -strlen($needle)) === $needle;
        }

        return false;
    }

    /**
     * Returns true if $haystack ends with $needle (case-insensitive)
     *
     *     Str::endsWith('foobar', 'bar');  // returns true
     *     Str::endsWith('foobar', 'baz');  // returns false
     *     Str::endsWith('foobar', 'BAR');  // returns true
     *
     * @param  string  $haystack  str  the string to search
     * @param  string  $needle    str  the substring to search for

     * @return  bool
     *
     * @see  self::iEndsWith()  case-sensitive version
     */
    public static function iEndsWith(string $haystack, string $needle): bool
    {
        return self::endsWith(strtolower($haystack), strtolower($needle));
    }

    /**
     * Alias for the isBool() method
     *
     * @see  self::isBool()
     */
    // phpcs:ignore PSR1.Methods.CamelCapsMethodName.NotCamelCaps  -- to match PHP's native method
    public static function is_bool($string): bool
    {
        return self::isBool($string);
    }

    /**
     * Returns true if $string is a bool string
     *
     * I'll return true if $string is a bool string like 'true', 'false', 'yes', 'no',
     * 'on' or 'off'. Keep in mind, I only handle strings. I will return false if you
     * test an actual bool value (because it's not a string).
     *
     *     is_bool(true);        // returns true
     *     Str::is_bool(true);   // returns false
     *
     *     is_bool('true');      // returns false
     *     Str::isBool('true');  // returns true
     *
     *     is_bool('yes');       // returns false
     *     Str::isBool('yes');   // returns true
     *
     * @param  string  $string  the string to test
     *
     * @return  bool
     */
    public static function isBool($string): bool
    {
        return is_string($string)
            && in_array(strtolower($string), ['true', 'false', 'yes', 'no', 'on', 'off']);
    }

    /**
     * Returns true if $haystack starts with $needle (case-insensitive)
     *
     * For example:
     *
     *     Str::iStartsWith('foobar', 'bar');  // returns false
     *     Str::iStartsWith('foobar', 'foo');  // returns true
     *     Str::iStartsWith('foobar', 'FOO');  // returns true
     *     Str::iStartsWith('', 'foobar');     // returns false
     *     Str::iStartsWith('foobar', '');     // returns false
     *
     * @param  string  $haystack  the case-insensitive string to search
     * @param  string  $needle    the case-insensitive substring to search for
     *
     * @return  bool  true if $haystack ends with $needle
     *
     * @see     self::startsWith()  case-sensitive version
     */
    public static function iStartsWith(string $haystack, string $needle): bool
    {
        return self::startsWith(strtolower($haystack), strtolower($needle));
    }

    /**
     * Returns a random string of $length that follows the charset $rules
     *
     * Oftetimes, standards (like PCI) require passwords with one upper-case letter, one
     * lower-case letter, one number, and one symbol. I can do that.
     *
     * For example:
     *
     *     $rules = ['upper' => 12];
     *     $a = Str::password(12, $rules);
     *
     *     $rules = ['lower' => 6, 'upper' => 6];
     *     $b = Str::password(12, $rules);
     *
     *     $rules = ['lower' => 4, 'upper' => 4, 'number' => 4];
     *     $c = Str::password(12, $rules);
     *
     *     echo $a;  // example 'KNVHYUIDGVDS'
     *     echo $b;  // example 'jNhGFkLekOfV'
     *     echo $c;  // example 'la9Uh7BH4Bc3'
     *
     * @param  int    $length  the length of the password (optional; if omitted,
     *    defaults to 8)
     * @param  int[]  $rules   an array of character counts indexed by charset name
     *    (possible charset names are 'lower', 'upper', 'number', 'alpha', and 'symbol')
     *    (optional; if omitted, defaults to ['lower' => 1, 'upper' => 1, 'number' => 1,
     *    'symbol' => 1])
     *
     * @return  string  the password
     *
     * @throws  \InvalidArgumentException  if a key in $rules is not a valid charset name
     * @throws  \InvalidArgumentException  if a value in $rules is not an integer
     * @throws  \InvalidArgumentException  if the number of required characters (as defined
     *    in the $rules array) exceeds the $length
     */
    public static function password(
        int $length = 8,
        array $rules = ['lower' => 1, 'upper' => 1, 'number' => 1, 'symbol' => 1]
    ): string {
        // if the number of required characters is GTE the desired length
        if (array_sum($rules) > $length) {
            throw new \InvalidArgumentException(
                "the number of required characters should be less than or equal to the length"
            );
        }

        // loop through the password's rules
        $password = '';
        foreach ($rules as $charset => $number) {
            $password .= self::rand($number, $charset);
        }

        // if any characters are missing, add them
        if ($length - strlen($password) > 0) {
            $password .= self::rand($length - strlen($password));
        }

        // shuffle the password
        $password = str_shuffle($password);

        return $password;
    }

    /**
     * Returns a random string
     *
     * For example:
     *
     *     echo Str::rand(8, 'alpha');              // example 'hbdrckso'
     *     echo Str::rand(8, ['lower', 'number']);  // example 'k987hb54'
     *     echo Str::rand(8, ['upper', 'symbol']);  // example 'HG!V*X]@'
     *
     * @param  int    $length    the length of the string to return
     * @param  mixed  $charsets  a string charset name or an array of charset names
     *    (possible values are are 'lower', 'upper', 'alpha' (a combination of 'upper'
     *    and 'lower'), 'number', and 'symbol') (optional; if omitted, defaults to
     *    ['alpha', 'number', 'symbol'])
     *
     * @return  string  a random string
     *
     * @throws  \InvalidArgumentException  if $charsets is not a string or array
     * @throws  \InvalidArgumentException  if a given $charset is not a valid charset
     */
    public static function rand(int $length, $charsets = ['alpha', 'number', 'symbol']): string
    {
        $rand = '';

        if (is_string($charsets)) {
            $charsets = (array)$charsets;
        }

        if (!is_array($charsets)) {
            throw new \InvalidArgumentException(
                "the second parameter, charsets, should be a string charset ".
                    "name or an array of charset names"
            );
        }

        // define the possible charsets
        $lower  = ['a', 'b', 'c', 'd', 'e', 'f', 'g', 'h', 'i', 'j', 'k', 'l',
            'm', 'n', 'o', 'p', 'q', 'r', 's', 't', 'u', 'v', 'w', 'x', 'y', 'z'];
        $upper  = ['A', 'B', 'C', 'D', 'E', 'F', 'G', 'H', 'I', 'J', 'K', 'L',
            'M', 'N', 'O', 'P', 'Q', 'R', 'S', 'T', 'U', 'V', 'W', 'X', 'Y', 'Z'];
        $number = ['0', '1', '2', '3', '4', '5', '6', '7', '8', '9'];
        $symbol = ['!', '@', '#', '*', '(', ')', '-', '_', '+', '=', '[', ']'];

        // create an array of possible chars
        $chars = [];
        foreach ($charsets as $charset) {
            if (isset($$charset)) {
                $chars = array_merge($chars, $$charset);
            } elseif ($charset === 'alpha') {
                $chars = array_merge($chars, $lower, $upper);
            } else {
                throw new \InvalidArgumentException(
                    __METHOD__." expects parameter two to be a string charset name or an array ".
                        "of charset names such as 'lower', 'upper', 'alpha', 'number', or 'symbol'"
                );
            }
        }

        // shuffle the chars
        shuffle($chars);

        // pick $length random chars
        for ($i = 0; $i < $length; ++$i) {
            $rand .= $chars[array_rand($chars)];
        }

        return $rand;
    }

    /**
     * Splits a string on the first alpha character
     *
     * I'll return an array with two parts. The first element is the string part before
     * the first alpha character, and the second part is everything after and including
     * the first alpha character.
     *
     * For example:
     *
     *     Str::splitOnFirstAlpha("123");        // returns ["123"]
     *     Str::splitOnFirstAlpha("abc");        // returns ["", "abc"]
     *     Str::splitOnFirstAlpha("123 abc");    // returns ["123", "abc"]
     *     Str::splitOnFirstAlpha("1 2 3 abc");  // returns ["1 2 3 4", "abc"]
     *
     * @param  string  $string  the string to split
     *
     * @return  string[]  an array
     *
     * @see  http://stackoverflow.com/a/18990341  FrankieTheKneeMan's answer to "Split
     *    string on first occurrence of a letter" on StackOverflow (version using Regex
     *    lookahead)
     */
    public static function splitOnFirstAlpha(string $string): array
    {
        $string = trim($string);

        if ($string === '') {
            return [];
        }

        return array_map('trim', preg_split('/(?=[a-zA-Z])/i', $string, 2));
    }

    /**
     * Returns true if $haystack starts with $needle (case-sensitive)
     *
     * For example:
     *
     *     Str::startsWith('foobar', 'bar');  // returns false
     *     Str::startsWith('foobar', 'foo');  // returns true
     *     Str::startsWith('foobar', 'FOO');  // returns false
     *     Str::startsWith('foobar', '');     // returns false
     *     Str::startsWith('', 'foobar');     // returns false
     *
     * @param  string  $haystack  the string to search
     * @param  string  $needle    the substring to search for
     *
     * @return  bool  true if $haystack starts with $needle
     *
     * @see  self::iStartsWith()  case-insensitive version
     * @see  http://stackoverflow.com/a/834355  MrHus' answer to "startsWith() and
     *    endsWith() functions in PHP" on StackOverflow
     */
    public static function startsWith(string $haystack, string $needle): bool
    {
        if (strlen($haystack) > 0 && strlen($needle) > 0) {
            return !strncmp($haystack, $needle, strlen($needle));
        }

        return false;
    }

    /**
     * Converts a php.ini-like byte notation shorthand to a number of bytes
     *
     * In the php.ini configuration file, byte values are sote in shorthand
     * notation (e.g., "8M"). PHP's native ini_get() function will return the
     * exact string stored in php.ini and not its integer equivalent. I will
     * return the integer equivalent.
     *
     * For example:
     *
     *     Str::strtobytes('1K');  // returns 1024
     *     Str::strtobytes('1M');  // returns 1048576
     *     Str::strtobytes('1G');  // returns 1073741824
     *
     * @param  string  $string  the string to convert
     *
     * @return  int|float  the number of bytes
     *
     * @throws  \InvalidArgumentException  if $string does not end in 'k', 'm', or 'g'
     *
     * @see     http://www.php.net/manual/en/function.ini-get.php  ini_get() man page
     */
    public static function strtobytes(string $string): int
    {
        $string = trim($string);

        $last = strtolower($string[strlen($string) - 1]);

        $value = 1;
        switch ($last) {
            case 'g':
                $value *= 1024;
                // no break

            case 'm':
                $value *= 1024;
                // no break

            case 'k':
                $value *= 1024;
                break;

            default:
                throw new \InvalidArgumentException(
                    "the string should end in 'k', 'm', or 'g'"
                );
        }

        return $value;
    }

    /**
     * Returns a string in camel case
     *
     * For example:
     *
     *     Str::strtocamelcase('Hello world');   // returns "helloWorld"
     *     Str::strtocamelcase('H3LLO WORLD!');  // returns "helloWorld"
     *     Str::strtocamelcase('hello_world');   // returns "helloWorld"
     *
     * @param  string  $string  the string to camel-case
     *
     * @return  string  the camel-cased string
     */
    public static function strtocamelcase(string $string): string
    {
        $string = trim($string);

        // replace underscores ("_") and hyphens ("-") with spaces (" ")
        $string = str_replace(['-', '_'], ' ', $string);

        // lower-case everything
        $string = strtolower($string);

        // capitalize each word
        $string = ucwords($string);

        // remove spaces
        $string = str_replace(' ', '', $string);

        // lower-case the first word
        $string = lcfirst($string);

        // remove any non-alphanumeric characters
        $string = preg_replace("#[^a-zA-Z0-9]+#", '', $string);

        return $string;
    }

    /**
     * Truncates $string to a preferred length
     *
     *     Str::truncate('Lorem ipsum inum', 8);             // returns 'Lorem ipsum...'
     *     Str::truncate('Lorem ipsum inum', 8, '');         // returns 'Lorem ip...'
     *     Str::truncate('Lorem ipsum inum', 8, ' ', ' >');  // returns 'Lorem ipsum >'
     *
     * @param  string  $str    the string to truncate
     * @param  int     $limit  the string's max length
     * @param  string  $break  the break character (to truncate at exact length set to
     *    empty string or null) (if the break character does not exist in the string,
     *    the string will be truncated at limit) (optional; if omitted, defaults to ' ')
     * @param  string  $pad    the padding to add to end of string (optional; if
     *    omitted, defaults to '...')
     *
     * @return  string  the truncated string
     *
     * @see     http://blog.justin.kelly.org.au/php-truncate/  The original function
     *    from "Best PHP Truncate Function" posted 6/27/12 on "Justin Kelly - various
     *    ramblings" (edited to find closest break *before* limit and truncate string
     *    exactly if break does not exist)
     */
    public static function truncate(
        string $string,
        int $limit,
        ?string $break = ' ',
        ?string $pad = '...'
    ): string {
        // if the string doesn't need truncating, short-circuit
        if (strlen($string) <= $limit) {
            return $string;
        }

        // truncate the string at the limit
        $truncated = substr($string, 0, $limit);

        // if a break character is defined and it exists in the truncated string
        if ($break && strpos($truncated, $break)) {
            $truncated = substr($truncated, 0, strrpos($truncated, $break));
        }

        // if a pad exists, use it
        if ($pad) {
            $truncated .= $pad;
        }

        return $truncated;
    }
}
